kratos/utilities/sensitivity_builder.cpp

```diff
@@ -0,0 +1,627 @@
+            << " incompatible with SensitivityMatrix.size1(): "
+            << SensitivityMatrix.size1() << ". Variable: " << rVariable << std::endl;
+        noalias(LocalSensitivity) = prod(SensitivityMatrix, AdjointVector) + PartialSensitivity;
+        return *this;
+        KRATOS_CATCH("");
+    }
+};
+
+/**
+ * @brief Call a function on each container element in a parallel loop
+ *
+ * Any data associated with the function (e.g., functors or capture by value)
+ * is copied to each thread.
+ */
+template <typename TContainer, typename TCallable>
+void ParallelForEach(TContainer& rContainer, TCallable&& Fun)
+{
+#pragma omp parallel
+    {
+        auto local_fun{Fun};
+#pragma omp for
+        for (int i = 0; i < static_cast<int>(rContainer.size()); ++i)
+        {
+            std::forward<TCallable>(local_fun)(*(rContainer.begin() + i));
+        }
+    }
+}
+
+/**
+ * @brief Contains the sensitivity design and output variables
+ *
+ * The design variable is passed to CalculateSensitivityMatrix()
+ * and CalculatePartialSensitivity() when calculating the local
+ * sensitivity contributions. The local sensitivities are assembled
+ * to the output variable.
+ *
+ * Example 1:
+ * SensitivityVariables<double> vars("THICKNESS");
+ * vars.pDesignVariable->Name(); // "THICKNESS"
+ * vars.pOutputVariable->Name(); // "THICKNESS_SENSITIVITY"
+ *
+ * Example 2:
+ * SensitivityVariables<double> vars("THICKNESS_SENSITIVITY");
+ * vars.pDesignVariable->Name(); // "THICKNESS_SENSITIVITY"
+ * vars.pOutputVariable->Name(); // "THICKNESS_SENSITIVITY"
+ */
+template <class TDataType>
+struct SensitivityVariables
+{
+    const Variable<TDataType>* pDesignVariable = nullptr;
+    const Variable<TDataType>* pOutputVariable = nullptr;
+
+    explicit SensitivityVariables(const std::string& rName)
+    {
+        KRATOS_TRY;
+        const std::string output_suffix = "_SENSITIVITY";
+        pDesignVariable = &KratosComponents<Variable<TDataType>>::Get(rName);
+        if (rName.size() > output_suffix.size() &&
+            std::equal(output_suffix.rbegin(), output_suffix.rend(), rName.rbegin()))
+        {
+            pOutputVariable = pDesignVariable;
+        }
+        else
+        {
+            pOutputVariable =
+                &KratosComponents<Variable<TDataType>>::Get(rName + output_suffix);
+        }
+        KRATOS_CATCH("");
+    }
+};
+
+template <typename TDataType, typename TContainer>
+void AssembleNodalSolutionStepContainerContributions(const SensitivityVariables<TDataType>& rVariables,
+                                                     TContainer& rContainer,
+                                                     AdjointResponseFunction& rResponseFunction,
+                                                     const ProcessInfo& rProcessInfo,
+                                                     double ScalingFactor)
+{
+    KRATOS_TRY;
+    LocalSensitivityBuilder builder;
```

Comment:
Just for me to understand: One builder is used inside each thread inside the parallel loop, right? Which is the related command? I know `#pragma omp parallel private(x)`) but i don't see something similar here?

Reply:
Instead of using `private(x)`, the builder is copied explicitly with `auto local_fun{Fun};` to each thread because openmp's `private` had some problems with the function parameter.